src/helper/Backend.cpp

```diff
@@ -70,6 +75,26 @@ namespace SDDM {
                         .arg(mainConfig.X11.UserAuthFile.get());
                 env.insert(QStringLiteral("XAUTHORITY"), value);
             }
+#if defined(Q_OS_FREEBSD)
+        /* get additional environment variables via setclassenvironment();
+            this needs to be done here instead of in UserSession::setupChildProcess
+            as the environment for execve() is prepared here */
+        login_cap_t *lc;
+
+        if (lc = login_getpwclass(pw)) {
+            setclassenvironment(lc, pw, 1);     /* path variables */
+            setclassenvironment(lc, pw, 0);     /* non-path variables */
+            login_close(lc);
+            if (lc = login_getuserclass(pw)) {
```

Comment:
Is it **only** possible to have a valid userclass setting, when there is already (also) a pwclass setting? If not, I'd structure this as:
```
if (lc = login_getXXclass(pw)) { ... ; login_close(lc); }
if (lc = login_getYYclass(pw)) { ... ; login_close(lc); }
```
Although then you'll need some kind of flag to indicate that `env.insert()` should be called (call it only if something was actually set).

Reply:
Re-routing original author's comment on this:

> This is modeled after [r357563 of FreeBSD/base](https://svnweb.freebsd.org/base/head/usr.bin/env/env.c?r1=357563&r2=357562&pathrev=357563) (line 173). I am not sure here, but it is probably the case that there is always a default pwclass configured so that would be found anyway. Also, I think it is doubtful to allow logging in only with user settings.

Reply:
In downstream r527832 of the patch in question this was changed; my original question regarded the construction

```
if ((lc = login_getpwclass(pw)) != NULL) {
    ...
    if ((lc = login_getuserclass(pw)) != NULL) {
        ...
    }
}
```

Note the nested `if()`s; I think it would be possible to have a valid userclass without a valid pwclass. The revised downstream patch follows the un-nested structure.

Reply:
>  I think it would be possible to have a valid userclass without a valid pwclass.

But as already noted, `/usr/bin/env` stops further processing if `login_getpwclass` returns NULL. I think this is because `login_getpwclass` falls back to `default` class if the `pw` points to a nonexistent record, so NULL returned means error state. I believe that nested structure should be preserved.